From abc77b36ac122f38c6f34c4cddcabff24b86b41e Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Wed, 10 Nov 2010 12:10:06 +0200
Subject: [PATCH 005/226] Fix strange initialisation code.

---
 src/map/pc.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/map/pc.c b/src/map/pc.c
index 11cd864..04a4062 100644
--- a/src/map/pc.c
+++ b/src/map/pc.c
@@ -5665,8 +5665,8 @@ int pc_resetlvl (struct map_session_data *sd, int type)
         sd->status.skill_point = 0;
         sd->status.base_level = 1;
         sd->status.job_level = 1;
-        sd->status.base_exp = sd->status.base_exp = 0;
-        sd->status.job_exp = sd->status.job_exp = 0;
+        sd->status.base_exp = 0;
+        sd->status.job_exp = 0;
         if (sd->status.option != 0)
             sd->status.option = 0;
 
-- 
2.1.0

