From 2f266d5a7416c3fcff2142022192509c76e3e624 Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Wed, 10 Nov 2010 23:51:02 +0200
Subject: [PATCH 006/226] Limit hair styles count to 31.

---
 src/char/char.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/char/char.c b/src/char/char.c
index fefc420..f526200 100644
--- a/src/char/char.c
+++ b/src/char/char.c
@@ -941,7 +941,7 @@ int make_new_char (int fd, unsigned char *dat)
 
     if (dat[24] + dat[25] + dat[26] + dat[27] + dat[28] + dat[29] != 5 * 6 ||   // stats
         dat[30] >= 9 ||         // slots (dat[30] can not be negativ)
-        dat[33] < 0 || dat[33] >= 20 || // hair style
+        dat[33] < 0 || dat[33] >= 30 || // hair style
         dat[31] >= 12)
     {                           // hair color (dat[31] can not be negativ)
         char_log
-- 
2.1.0

