From 2a364172ecc3fbfaa50da0491e96f53ec5392e5c Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Fri, 12 Nov 2010 02:39:05 +0200
Subject: [PATCH 009/226] Add one check to clif.c

---
 src/map/clif.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/map/clif.c b/src/map/clif.c
index 2eeee35..7eac704 100644
--- a/src/map/clif.c
+++ b/src/map/clif.c
@@ -143,6 +143,9 @@ char talkie_mes[80];
  */
 void clif_setip (char *ip)
 {
+    if (!ip)
+        return;
+
     memcpy (map_ip_str, ip, 16);
     map_ip = inet_addr (map_ip_str);
 }
-- 
2.1.0

