From fca5c975c1948e246c45dd8ed26c344815b83bfc Mon Sep 17 00:00:00 2001
From: "ea@ak64.ml" <ea@ak64.ml>
Date: Fri, 12 Nov 2010 02:07:56 +0000
Subject: [PATCH 011/226] add static compilation.

---
 make.defs        | 6 ++++--
 src/map/Makefile | 3 ++-
 2 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/make.defs b/make.defs
index b9bc3fb..0ca8fbf 100644
--- a/make.defs
+++ b/make.defs
@@ -4,7 +4,9 @@ ifeq ($(shell uname -m), x86_64)
 M32=-m32
 endif
 
-CC = gcc ${M32}
+STATIC=-static
+
+CC = gcc ${M32} ${STATIC}
 LEX=flex
 BISON=bison
 
@@ -22,7 +24,7 @@ OS_TYPE = -DCYGWIN
 CFLAGS = $(OPT) -DFD_SETSIZE=4096 -I../common $(PACKETDEF) $(OS_TYPE)
 else
 OS_TYPE =
-CFLAGS = $(OPT) -fstack-protector -Wno-pointer-sign -I../common $(PACKETDEF) $(OS_TYPE)
+CFLAGS = $(OPT) ${STATIC} -fstack-protector -Wno-pointer-sign -I../common $(PACKETDEF) $(OS_TYPE)
 endif
 
 MKDEF = CC="$(CC)" CFLAGS="$(CFLAGS)"
diff --git a/src/map/Makefile b/src/map/Makefile
index a1a4381..39769ca 100644
--- a/src/map/Makefile
+++ b/src/map/Makefile
@@ -8,13 +8,14 @@ obj:
 	mkdir obj
 
 COMMON_OBJ = ../common/core.o ../common/socket.o ../common/timer.o ../common/grfio.o ../common/db.o ../common/lock.o ../common/nullpo.o ../common/malloc.o ../common/mt_rand.o ../common/md5calc.o
+
 LIBS = -lz -lm
 
 map-server: obj/tmw.o obj/magic-interpreter-lexer.o obj/magic-interpreter-parser.o obj/magic-interpreter-base.o obj/magic-expr.o obj/magic-stmt.o obj/magic.o obj/map.o obj/chrif.o obj/clif.o obj/pc.o obj/npc.o obj/chat.o obj/path.o obj/itemdb.o obj/mob.o obj/script.o obj/storage.o obj/skill.o obj/skill-pools.o obj/atcommand.o obj/battle.o obj/intif.o obj/trade.o obj/party.o obj/guild.o $(COMMON_OBJ)
 	$(CC) -o ../../$@ $^ $(LIBS)
 
 obj/%.o: %.c
-	$(COMPILE.c) $(OUTPUT_OPTION) $<
+	$(COMPILE.c) ${STATIC} $(OUTPUT_OPTION) $<
 
 magic-interpreter-lexer.c: magic-interpreter.l
 	$(LEX) -o magic-interpreter-lexer.c magic-interpreter.l
-- 
2.1.0

