From 05491c988f10d5223bafdd63e096ac481d991085 Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Tue, 16 Nov 2010 21:34:13 +0200
Subject: [PATCH 025/226] Update readme file.

---
 README | 10 +++-------
 1 file changed, 3 insertions(+), 7 deletions(-)

diff --git a/README b/README
index ec80074..ba5eb12 100644
--- a/README
+++ b/README
@@ -1,8 +1,4 @@
-This is just eAthena. We store our scripts, item and monster databases, etc.
-in a separate git repository called tmw-eathena-data, available at:
+This is evol server code.
 
- http://gitorious.org/projects/tmw-eathena-data
-
-Note that this repository contains more than just the contents of this data
-directory. Either set up symbolic links or just copy the server executables
-into it.
+Client or server data you can find in other evol repositories here:
+    http://www.gitorious.org/evol
-- 
2.1.0

