From 757a63668925d90b5c71e08486370da02521ed06 Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Wed, 17 Nov 2010 21:46:37 +0200
Subject: [PATCH 028/226] Add checks to path.c

---
 src/map/path.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/map/path.c b/src/map/path.c
index a6e7535..99449c3 100644
--- a/src/map/path.c
+++ b/src/map/path.c
@@ -271,14 +271,14 @@ int path_blownpos (int m, int x0, int y0, int dx, int dy, int count)
 int path_search (struct walkpath_data *wpd, int m, int x0, int y0, int x1,
                  int y1, int flag)
 {
+    nullpo_retr (0, wpd);
+
     int  heap[MAX_HEAP + 1];
     struct tmp_path tp[MAX_WALKPATH * MAX_WALKPATH];
     int  i, rp, x, y;
     struct map_data *md;
     int  dx, dy;
 
-    nullpo_retr (0, wpd);
-
     if (!map[m].gat)
         return -1;
     md = &map[m];
-- 
2.1.0

