From d927f7a7bc27fb98706028a54f86df76c1899aea Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Mon, 22 Nov 2010 22:04:53 +0200
Subject: [PATCH 035/226] Allow exchange between cart and storage only for
 carton role.

---
 src/map/clif.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/map/clif.c b/src/map/clif.c
index 98e4639..68aa54a 100644
--- a/src/map/clif.c
+++ b/src/map/clif.c
@@ -8635,7 +8635,7 @@ void clif_parse_MoveToKafraFromCart (int fd, struct map_session_data *sd)
     nullpo_retv (sd);
 
     if ((sd->npc_id != 0 && !sd->npc_flags.storage) || sd->trade_partner != 0
-        || !sd->state.storage_flag)
+        || !sd->state.storage_flag || !pc_iscarton (sd))
         return;
     if (sd->state.storage_flag == 1)
         storage_storageaddfromcart (sd, RFIFOW (fd, 2) - 2, RFIFOL (fd, 4));
@@ -8653,7 +8653,7 @@ void clif_parse_MoveFromKafraToCart (int fd, struct map_session_data *sd)
     nullpo_retv (sd);
 
     if ((sd->npc_id != 0 && !sd->npc_flags.storage) || sd->trade_partner != 0
-        || !sd->state.storage_flag)
+        || !sd->state.storage_flag || !pc_iscarton (sd))
         return;
     if (sd->state.storage_flag == 1)
         storage_storagegettocart (sd, RFIFOW (fd, 2) - 1, RFIFOL (fd, 4));
-- 
2.1.0

