From 24177214fc5c2200d9aa6e92babeb51755f11862 Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Wed, 24 Nov 2010 03:18:51 +0200
Subject: [PATCH 041/226] Add checks to storage.c

---
 src/map/storage.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/map/storage.c b/src/map/storage.c
index 075f396..d6c60e5 100644
--- a/src/map/storage.c
+++ b/src/map/storage.c
@@ -91,6 +91,9 @@ void do_final_storage (void)    // by [MC Cameri]
 
 static int storage_reconnect_sub (void *key __attribute__ ((unused)), void *data, va_list ap)
 {                               //Parses storage and saves 'dirty' ones upon reconnect. [Skotlex]
+    nullpo_retr (0, ap);
+    nullpo_retr (0, data);
+
     int  type = va_arg (ap, int);
     if (type)
     {                           //Guild Storage
@@ -186,6 +189,9 @@ static int storage_additem (struct map_session_data *sd, struct storage *stor,
     struct item_data *data;
     int  i;
 
+    nullpo_retr (1, stor);
+    nullpo_retr (1, item_data);
+
     if (item_data->nameid <= 0 || amount <= 0)
         return 1;
 
@@ -229,6 +235,10 @@ static int storage_delitem (struct map_session_data *sd, struct storage *stor,
                             int n, int amount)
 {
 
+    nullpo_retr (1, stor);
+    if (n < 0 || n >= MAX_STORAGE)
+        return 1;
+
     if (stor->storage_[n].nameid == 0 || stor->storage_[n].amount < amount)
         return 1;
 
@@ -422,6 +432,8 @@ int storage_storage_quit (struct map_session_data *sd)
 
 void storage_storage_dirty (struct map_session_data *sd)
 {
+    nullpo_retv (sd);
+
     struct storage *stor;
 
     stor = account2storage2 (sd->status.account_id);
@@ -594,6 +606,9 @@ int guild_storage_delitem (struct map_session_data *sd,
     nullpo_retr (1, sd);
     nullpo_retr (1, stor);
 
+    if (n < 0 || n >= MAX_GUILD_STORAGE)
+        return 1;
+
     if (stor->storage_[n].nameid == 0 || stor->storage_[n].amount < amount)
         return 1;
 
-- 
2.1.0

