From 2891fd8ba055ddc2c2cc1f49bd3aaf805092f53f Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Sat, 27 Nov 2010 01:38:05 +0200
Subject: [PATCH 049/226] Increase hair colors limit to 30.

---
 src/char/char.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/char/char.c b/src/char/char.c
index d7287c4..fc13651 100644
--- a/src/char/char.c
+++ b/src/char/char.c
@@ -983,7 +983,7 @@ int make_new_char (int fd, unsigned char *dat)
     if (dat[24] + dat[25] + dat[26] + dat[27] + dat[28] + dat[29] != 5 * 6 ||   // stats
         dat[30] >= 9 ||         // slots (dat[30] can not be negativ)
         /*dat[33] < 0 ||*/ dat[33] >= 30 || // hair style
-        dat[31] >= 12)
+        dat[31] >= 30)
     {                           // hair color (dat[31] can not be negativ)
         char_log
             ("Make new char error (invalid values): (connection #%d, account: %d) slot %d, name: %s, stats: %d+%d+%d+%d+%d+%d=%d, hair: %d, hair color: %d"
-- 
2.1.0

