From 55eecc37cf557fa028fd6c2af84a4e77077ed242 Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Thu, 2 Dec 2010 02:52:34 +0200
Subject: [PATCH 068/226] Fix crash in mob.c if monster have mutations.

---
 src/map/mob.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/map/mob.c b/src/map/mob.c
index 3d1aa5a..472c395 100644
--- a/src/map/mob.c
+++ b/src/map/mob.c
@@ -325,6 +325,9 @@ static void mob_init (struct mob_data *md)
     md->stats[MOB_SPEED] = mob_db[class].speed;
     md->stats[MOB_XP_BONUS] = MOB_XP_BONUS_BASE;
 
+    if (!(mutation_power >> 1))
+        return;
+
     for (i = 0; i < mutations_nr; i++)
     {
         int  stat_nr = MRAND (MOB_XP_BONUS + 1);
@@ -4788,6 +4791,8 @@ static int mob_readdb (void)
                     atoi (str[48 + i * 2]) * battle_config.mvp_item_rate /
                     100;
             }
+            // skip two positions
+
             mob_db[class].mutations_nr = atoi (str[55]);
             mob_db[class].mutation_power = atoi (str[56]);
 
-- 
2.1.0

