From c76e06f7acbf355aa6589f30c590103b15b2370b Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Fri, 7 Jan 2011 01:37:45 +0200
Subject: [PATCH 172/226] Add missing checks to skill-pools.c

---
 src/map/skill-pools.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/map/skill-pools.c b/src/map/skill-pools.c
index 9ffc2bd..31c0238 100644
--- a/src/map/skill-pools.c
+++ b/src/map/skill-pools.c
@@ -49,6 +49,9 @@ char *skill_name (int skill)
 
 int skill_pool (struct map_session_data *sd, int *skills)
 {
+    if (!sd)
+        return 0;
+
     int  i, count = 0;
 
     for (i = 0; count < MAX_SKILL_POOL && i < skill_pool_skills_size; i++)
@@ -72,7 +75,7 @@ void skill_pool_empty (struct map_session_data *sd)
     if (!sd)
         return;
 
-    for (i = 0; i < skill_pool_skills_size; i++)
+    for (i = 0; i < skill_pool_skills_size && i < MAX_SKILL; i++)
     {
         int  skill_id = skill_pool_skills[i];
         sd->status.skill[skill_id].flags = 0;
-- 
2.1.0

