From ea012e9f918929573ac7b2138cd69a5d7d74e739 Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Mon, 10 Jan 2011 04:07:19 +0200
Subject: [PATCH 180/226] Limit gm changing skill commands to valid skills and
 levels.

---
 src/map/atcommand.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/map/atcommand.c b/src/map/atcommand.c
index 6f20f2b..a3f8df6 100644
--- a/src/map/atcommand.c
+++ b/src/map/atcommand.c
@@ -8916,9 +8916,12 @@ atcommand_magic_info (const int fd, struct map_session_data *sd,
 
 static void set_skill (struct map_session_data *sd, int i, int level)
 {
-    if (!sd || i < 0 || i >= MAX_SKILL)
+    if (!sd || i < 0 || i >= MAX_SKILL || skill_db[i].max == 0)
         return;
 
+    if (level > skill_db[i].max)
+        level = skill_db[i].max;
+
     sd->status.skill[i].id = level ? i : 0;
     sd->status.skill[i].lv = level;
 }
-- 
2.1.0

