From 8f43da1495993f074b44a4f57f94087b7a44bac7 Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Wed, 2 Feb 2011 00:36:43 +0200
Subject: [PATCH 215/226] Add script command settempnpcclass.

---
 src/map/script.c | 33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/src/map/script.c b/src/map/script.c
index 15fac90..5fe9990 100644
--- a/src/map/script.c
+++ b/src/map/script.c
@@ -457,6 +457,7 @@ int  buildin_atoi (struct script_state *st);
 int  buildin_rid2name (struct script_state *st);
 int  buildin_setnpcclass (struct script_state *st); // [4144]
 int  buildin_getnpcclass (struct script_state *st); // [4144]
+int  buildin_settempnpcclass (struct script_state *st); // [4144]
 int  buildin_l (struct script_state *st); // [4144]
 int  buildin_setlang (struct script_state *st); // [4144]
 int  buildin_getlang (struct script_state *st); // [4144]
@@ -970,6 +971,8 @@ struct
     {
     buildin_getnpcclass, "getnpcclass", "*"}, // [4144]
     {
+    buildin_settempnpcclass, "settempnpcclass", "*"}, // [4144]
+    {
     buildin_l, "l", "s"}, // [4144]
     {
     buildin_getlang, "getlang", "*"}, // [4144]
@@ -8984,6 +8987,36 @@ BUILDIN_FUNC(setnpcclass)
     return 0;
 }
 
+BUILDIN_FUNC(settempnpcclass)
+{
+    int  newsprite;
+    int  oldsprite;
+    struct npc_data *nd = 0;
+
+    if (script_hasdata(st, 3))
+    {
+        nd = npc_name2id (script_getstr(st, 2));
+        newsprite = script_getnum(st, 3);
+    }
+    else if (script_hasdata(st, 2))
+    {
+        if (!st->oid)
+            return 1;
+
+        nd = (struct npc_data *) map_id2bl (st->oid);
+        newsprite = script_getnum(st, 2);
+    }
+    if (!nd)
+        return 1;
+
+    oldsprite = nd->class;
+    nd->class = newsprite;
+    npc_enable (nd->name, 0);
+    npc_enable (nd->name, 1);
+    nd->class = oldsprite;
+    return 0;
+}
+
 BUILDIN_FUNC(getnpcclass)
 {
     struct npc_data *nd = 0;
-- 
2.1.0

