From ac7d12a76865ef0c87c03128812fe399a520f68a Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Thu, 3 Feb 2011 17:04:18 +0200
Subject: [PATCH 219/226] Add script command g. It show message depend on
 player gender.

---
 src/map/script.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/map/script.c b/src/map/script.c
index 7ac2811..6873ca2 100644
--- a/src/map/script.c
+++ b/src/map/script.c
@@ -466,6 +466,7 @@ int  buildin_seta (struct script_state *st); // [4144]
 int  buildin_geta2 (struct script_state *st); // [4144]
 int  buildin_seta2 (struct script_state *st); // [4144]
 int  buildin_getclientversion (struct script_state *st); // [4144]
+int  buildin_g (struct script_state *st); // [4144]
 
 void push_val (struct script_stack *stack, int type, int val);
 int  run_func (struct script_state *st);
@@ -991,6 +992,8 @@ struct
     {
     buildin_getclientversion, "getclientversion", "*"}, // [4144]
     {
+    buildin_g, "g", "ss"}, // [4144]
+    {
 NULL, NULL, NULL},};
 
 //int  buildin_message (struct script_state *st); // [MouseJstr]
@@ -9276,6 +9279,21 @@ BUILDIN_FUNC(getclientversion)
     return 0;
 }
 
+
+BUILDIN_FUNC(g)
+{
+    TBL_PC *sd = NULL;
+
+    sd = script_rid2sd(st);
+    if (sd == NULL || !sd->sex)
+        script_pushstr(st, (unsigned char *)strdup(script_getstr(st, 2)));
+    else
+        script_pushstr(st, (unsigned char *)strdup(script_getstr(st, 3)));
+
+    return 0;
+}
+
+
 //
 // 実行部main 
 //
-- 
2.1.0

