From 814f08919ac12b6c2a63a92c42a55569e6b2994d Mon Sep 17 00:00:00 2001
From: Andrei Karas <akaras@inbox.ru>
Date: Mon, 7 Feb 2011 04:27:54 +0200
Subject: [PATCH 225/226] Ignoring chat spam check for incorrect magic
 commands.

---
 src/map/clif.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/map/clif.c b/src/map/clif.c
index ee64515..961f46d 100644
--- a/src/map/clif.c
+++ b/src/map/clif.c
@@ -7465,6 +7465,13 @@ void clif_parse_GlobalMessage (int fd, struct map_session_data *sd)
 
     if (!magic_message (sd, buf, msg_len, 1))
     {
+        /* ignore incorrect magic commands */
+        if (message && strlen(message) > 3 && *message == '#' && *(message + 1) != '#')
+        {
+            free (buf);
+            return;
+        }
+
         /* Don't send chat that results in an automatic ban. */
         if (tmw_CheckChatSpam (sd, message))
         {
-- 
2.1.0

